NFT/src/main.sw

```diff
@@ -0,0 +1,368 @@
+contract;
+
+dep contract_abi;
+dep data_structures;
+dep errors;
+dep events;
+
+use contract_abi::NFT;
+use data_structures::MetaData;
+use errors::{AccessError, ApprovalError, InitError, InputError};
+use events::{ApprovalEvent, BurnEvent, MintEvent, OperatorEvent, TransferEvent};
+
+use std::{
+    assert::require,
+    chain::auth::{AuthError, msg_sender},
+    hash::sha256,
+    identity::Identity,
+    logging::log,
+    option::Option,
+    result::Result,
+    storage::StorageMap,
+    vec::Vec,
+};
+
+storage {
+    /// Determines if only the `admin` is allowed to call the mint function.
+    /// This is only set on the initalization of the contract.
+    access_control: bool,
+
+    /// Stores an `Option` of the `Identity` that has permission to mint if `access_control` is set to true.
+    /// Will store `None` if this contract does not have `access_control` set.
+    /// Only the `admin` is allowed to change the `admin` of the contract.
+    admin: Option<Identity>,
+
+    /// Used of O(1) lookup of the number of tokens owned by each `Identity`.
+    /// This increments or decrements when minting, transfering ownership, and burning tokens.
+    /// Map(Identity => balance)
+    balances: StorageMap<Identity,
+    u64>, /// Stores the `Metadata` for each token based on the token's `u64` id
+    /// Map(token_id => Metadata)
+    meta_data: StorageMap<u64,
+    Option<MetaData>>, /// Maps a `b256` hash of the (owner, operator) identities and stores whether the
+    /// operator is allowed to transfer ALL tokens on the owner's behalf.
+    /// Map(hash => approved)
+    operator_approval: StorageMap<b256,
+    bool>, /// The total number of tokens that have been minted.
+    /// This should only be incremented.
+    token_count: u64,
+
+    /// The total supply tokens that can be minted. Can only be set on the initalization of the
+    /// contract.
+    token_supply: u64,
+}
+
+impl NFT for Contract {
+    /// Constructor for the NFT. Calling this function will instantiate the `total_supply`, the `admin`
```

Comment:
Did we decide on putting docstrings on implementations? The issue with this is that a downstream contract of this contract only has access to the ABI file, not the implementation (of which there may be many). They wouldn't actually be able to see these docstrings.

As such, documentation should go on the ABI declaration.

Reply:
Interesting. I was only considering the implementation since that is what you are reading and it makes it easier to match up the docs with the code since you can see it right there.

At the same time I dislike the idea of not having docs on the functions themselves. It feels like it might become disjoint if not carefully maintained. It's also not a good idea to duplicate in the abi and impl.

I will make adjustments so that the docs are on the ABI instead when I get around to upgrading the contracts in some issues - some of which are already as drafts.

Reply:
In general, I agree that it makes more sense to add docstrings to the ABI declaration (even though I like to prefer reading it right on the implementation in practice), and also that we don't want duplicated docs for maintenance reasons. 
On the other hand, the implementation should add additional documentation regarding details specific to the implementation ( i.e: additional cases which may cause a revert).